### ISSUE TITLE
[SEP 6] Implement the /deposit endpoint

### ISSUE THREAD

**Issue by accordeiro** [stellar/stellar-anchor-server#14]

**Comment by accordeiro** [comment]
@tomerweller @tomquisel do the steps described above make sense to you?

There are a few open questions on my end:

1. I'm not sure what the best way for accomplishing the traceability on item 4 is, since this will heavily vary on the deposit method (e.g. we can use a tag when sending Ripple, but we'd likely need a different way for providing an identifier through a SWIFT deposit). 
I guess if we were making a specific implementation we could hardcode the format for each, but since  this is generic, what would be the best approach? Maybe adding a small templating language in which you could add `%TRANSACTION_ID%` to either the `how` or in the  `extra_info` field for a given `DepositInfo` entity?

2. Where should we draw funds from in order to address the edge cases on item 6? Should we use Friendbot for funding an "escrow" account, and use it for sending the `2.01 XLM` to newly created accounts for now?

Please let me know if I need to clarify any of these questions further :)

**Comment by tomquisel** [comment]
This makes sense to me!

For your questions:

1. I might just implement one or two example ways, and do it so it's easily pluggable, rather than trying to do templating. Templating is a little limited since I think not every deposit method will match the template. You could have a separate function that generates the deposit response for each type, and then generate an example function for BTC and an example function for SWIFT? I think Tempo has a SWIFT response already, so you could mimic that. The example function wouldn't actually do any work on the BTC or SWIFT networks, it's just generate a deposit response in the correct format with random made-up addresses.

2. I like the idea of using friendbot to fund an account that's then used to fund user accounts. Maybe it's not quite an escrow account, though? It's more of a creating or funding account.

Also, for this idea:
> 5. Create a cron job that goes over all deposit Transaction entities with pending_external status, and checks if the external transaction has been fulfilled (this checking can be done through a mock function for now).

I might instead expose an internal API that you can hit to tell the anchor server that an external transaction has been filled. That makes demos easier, because you can manually hit that endpoint to move the process along, rather than needing to change some state for the cron checker to detect. It's also a more responsive architecture, since it's push-based rather than polling-based.

Let me know your thoughts, I might not have considered everything.

**Comment by accordeiro** [comment]
This is very helpful, thanks! Following up:

1. That's a great suggestion – I think I was trying to be too generic by allowing the admin to add any random DepositType, and that probably wouldn't be how Anchors would implement this anyways. I think I'll replace that with an enum field, that you can basically select a couple of options that map to a function that generates the deposit response, as you've suggested.

2. Yeah, it's definitely not an escrow account. `Funding account` would definitely be a more suitable name for it :)

5. I love the idea of creating a webhook to let the server know whenever an external transaction has ben filled. @tomquisel how would you proceed about the other tasks that follow the receipt of the external transaction, such as account funding, trustline creation and sending the assets to the account? Would you do that synchronously whenever that webhook is called, and before sending the response to the webhook?

**Comment by tomquisel** [comment]
Cool! For 3, I'd probably make the external transaction webhook async since it could take a while for the user to add a trustline once their account has been created.

**Comment by accordeiro** [comment]
Yep, that makes sense – I think I've got a clearer roadmap now! I've updated the list of tasks above to reflect our discussion – please let me know if anything seems out of place.

Thanks! :)
